exporter/signalfxexporter/translation/converter.go

```diff
@@ -475,5 +467,113 @@ func (c *MetricsConverter) ConvertDimension(dim string) string {
 	if c.metricTranslator != nil {
 		res = c.metricTranslator.translateDimension(dim)
 	}
-	return filterKeyChars(res, c.nonAlphanumericDimChars)
+	return filterKeyChars(res, c.datapointValidator.nonAlphanumericDimChars)
+}
+
+// Values obtained from https://dev.splunk.com/observability/docs/datamodel/ingest#Criteria-for-metric-and-dimension-names-and-values
+const (
+	maxMetricNameLength     = 256
+	maxDimensionNameLength  = 128
+	maxDimensionValueLength = 256
+)
+
+type datapointValidator struct {
+	logger                  *zap.Logger
+	nonAlphanumericDimChars string
+}
+
+func newDatapointValidator(logger *zap.Logger, nonAlphanumericDimChars string) *datapointValidator {
+	return &datapointValidator{logger: createSampledLogger(logger), nonAlphanumericDimChars: nonAlphanumericDimChars}
+}
+
+// sanitizeDataPoints sanitizes datapoints prior to dispatching them to the backend.
+// Datapoints that do not conform to the requirements are removed. This method drops
+// datapoints with metric name greater than 256 characters.
+func (dpv *datapointValidator) sanitizeDataPoints(dps []*sfxpb.DataPoint) []*sfxpb.DataPoint {
+	resultDatapointsLen := 0
+	for dpIndex, dp := range dps {
+		if dpv.isValidMetricName(dp.Metric) {
+			dp.Dimensions = dpv.sanitizeDimensions(dp.Dimensions)
+			if resultDatapointsLen < dpIndex {
+				dps[resultDatapointsLen] = dp
+			}
+			resultDatapointsLen++
+		}
+	}
+
+	// Trim datapoints slice to account for any removed datapoints.
+	return dps[:resultDatapointsLen]
+}
+
+// sanitizeDimensions replaces all characters unsupported by SignalFx backend
+// in metric label keys and with "_" and drops dimensions when the key is greater
+// than 128 characters or when value is greater than 256 characters in length.
+func (dpv *datapointValidator) sanitizeDimensions(dimensions []*sfxpb.Dimension) []*sfxpb.Dimension {
+	resultDimensionsLen := 0
+	for dimensionIndex, d := range dimensions {
+		if dpv.isValidDimension(d) {
+			d.Key = filterKeyChars(d.Key, dpv.nonAlphanumericDimChars)
+			if resultDimensionsLen < dimensionIndex {
+				dimensions[resultDimensionsLen] = d
+			}
+			resultDimensionsLen++
+		}
+	}
+
+	// Trim dimensions slice to account for any removed dimensions.
+	return dimensions[:resultDimensionsLen]
+}
+
+func (dpv *datapointValidator) isValidMetricName(name string) bool {
+	if len(name) > maxMetricNameLength {
+		dpv.logger.Warn("dropping datapoint",
+			zap.String("reason", "metric name too long"),
+			zap.String("metric_name", name))
+		return false
+	}
+	return true
+}
+
+func (dpv *datapointValidator) isValidDimension(dimension *sfxpb.Dimension) bool {
+	return dpv.isValidDimensionName(dimension.Key) && dpv.isValidDimensionValue(dimension.Value)
+}
+
+func (dpv *datapointValidator) isValidDimensionName(name string) bool {
+	if len(name) > maxDimensionNameLength {
+		dpv.logger.Warn("dropping dimension",
+			zap.String("reason", "dimension name too long"),
```

Comment:
Instead of saying "too long" can we specifically say how long it is and what is our limit? It will help when troubleshooting.

Reply:
Yes, I think that makes sense. Updated per suggestion.